Introduce APIs to support multiple ClusterCIDRs

sarveshr7 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:
This PR implements APIs from https://github.com/kubernetes/enhancements/issues/2593

API Changes:

- Introduces networking.k8s.io/v1alpha1 API group
- Introduces ClusterCIDR type in networking.k8s.io/v1alpha1 API group

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
NONE

#### Special notes for your reviewer:
https://github.com/kubernetes/kubernetes/pull/108290 was already approved and merged in 1.24 , however was reverted by https://github.com/kubernetes/kubernetes/pull/109436 as it was merged post the 1.24 release cut deadline. Rebasing the old PR and planning to get it in for 1.25

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduce a v1alpha1 networking API for ClusterCIDR
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]:  https://github.com/kubernetes/enhancements/pull/2593
```

sarveshr7 commented:
/cc @aojea @thockin @rahulkjoshi

rahulkjoshi commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,81 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+
+func validatePerNodeHostBits(configCIDR string, perNodeHostBits, maxMaskSize int32, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	minPerNodeHostBits := int32(4)
```
We should document this limit in the field comment above.

sarveshr7 replied:
Done

rahulkjoshi commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -1982,3 +1982,187 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+
+	validNodeSelector := makeNodeSelector("foo", api.NodeSelectorOpIn, []string{"bar"})
+
+	successCases := map[string]*networking.ClusterCIDRConfig{
```
nit: please ignore if the rest of this file is written this way.

It feels weird to define test cases by modifying fields in a valid config rather than constructing the object you wish to test.

sarveshr7 replied:
Yes, the entire file is written this way, I followed the pattern looking at the `NetworkPolicy` and `IngressClass` tests

aojea replied:
agree it is weird ... and hard to read the tests :shrug:

sarveshr7 replied:
Done

rahulkjoshi commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -1982,3 +1982,187 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+		"valid IPv6 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv4CIDR(8, "")),
+		"valid IPv4 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv6CIDR(8, "")),
+		"valid DualStack ClusterCIDRConfig with no NodeSelector": makeClusterCIDRConfigCustom(setNodeSelector(nil)),
+		"valid NodeSelector":                                     makeClusterCIDRConfigCustom(setNodeSelector(validNodeSelector)),
```
nit: isn't this equivalent to passing `makeValidClusterCIDRConfig()`

sarveshr7 replied:
Ack. That's true, removed the case as it is already tested

rahulkjoshi commented on pkg/printers/internalversion/printers.go:
```diff
@@ -2624,6 +2637,57 @@ func printPriorityLevelConfigurationList(list *flowcontrol.PriorityLevelConfigur
+	row.Cells = append(row.Cells, obj.Name, fmt.Sprint(obj.Spec.PerNodeHostBits), ipv4, ipv6, translateTimestampSince(obj.CreationTimestamp))
+	if options.Wide {
+		nodeSelector := "<none>"
+		if obj.Spec.NodeSelector != nil {
```
can you use `metav1.FormatLabelSelector` (or some such equivalent) rather than defining our own? Feels like this should be a utility.

sarveshr7 replied:
I couldn't find a method for NodeSelector, should this be moved to metav1?

rahulkjoshi commented on pkg/registry/networking/clustercidrconfig/strategy.go:
```diff
@@ -0,0 +1,84 @@
+// PrepareForCreate clears the status of a ClusterCIDRConfig before creation.
+func (clusterCIDRConfigStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {}
+
+// PrepareForUpdate clears fields that are not allowed to be set by end users on update.
```
Is this function supposed to be empty -- the comment claims it's clearing fields?

sarveshr7 replied:
Yes, supposed to be empty, removed the comment

rahulkjoshi commented on pkg/apis/networking/types.go:
```diff
@@ -583,3 +583,61 @@ type ServiceBackendPort struct {
+	// address into 24 bits for the network portion and 8 bits for the host portion.
+	// For a /24 mask for IPv4 or a /120 for IPv6, configure PerNodeHostBits=8
+	// This field is immutable.
+	// +optional
```
Why make this optional? We know it will return an error if called without specifying `PerNodeHostBits`.

sarveshr7 replied:
I agree, made this required

sarveshr7 commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -1982,3 +1982,187 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+		"valid IPv4 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv6CIDR(8, "")),
+		"valid DualStack ClusterCIDRConfig with no NodeSelector": makeClusterCIDRConfigCustom(setNodeSelector(nil)),
+		"valid NodeSelector":                                     makeClusterCIDRConfigCustom(setNodeSelector(validNodeSelector)),
+	}
```
Done

bowei commented:
/ok-to-test

sarveshr7 commented:
/retest

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,85 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+	if spec.IPv4CIDR != "" {
+		if !netutils.IsIPv4CIDRString(spec.IPv4CIDR) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4CIDR"), spec.IPv4CIDR, "must be a valid IPv4 CIDR"))
+			return allErrs
```
we should not return, if someone sets wrong IPv4 and IPv6 CIDRs we should return both errors

sarveshr7 replied:
Done

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,85 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("perNodeHostBits"), perNodeHostBits, fmt.Sprintf("must be greater than %d", minPerNodeHostBits)))
+	}
+	if perNodeHostBits > maxPerNodeHostBits {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("perNodeHostBits"), perNodeHostBits, fmt.Sprintf("must be less than or equal to %d", maxPerNodeHostBits)))
```
I can't remember the discussion, equal means single node cluster, is that ok?

sarveshr7 replied:
Yes, that's the case where a Per Node CCC can be configured, for a single node in the cluster
e.g. 
```
ClusterCIDRConfig{
    PerNodeHostBits: 8
    IPv4CIDR: "192.168.0.0/24"
}
```

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -583,3 +583,62 @@ type ServiceBackendPort struct {
+	metav1.ObjectMeta
+
+	Spec   ClusterCIDRConfigSpec
+	Status ClusterCIDRConfigStatus
```
@thockin if we don't use it, can we drop it directly?

sarveshr7 replied:
Dropped

aojea commented on pkg/apis/networking/types.go:
```diff
@@ -583,3 +583,62 @@ type ServiceBackendPort struct {
+	IPv6CIDR string
+}
+
+// ClusterCIDRConfigStatus defines the observed state of ClusterCIDRConfig.
+type ClusterCIDRConfigStatus struct {
+}
```
https://github.com/kubernetes/kubernetes/pull/111123/files#r933135006

sarveshr7 replied:
Dropped

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,85 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+	return allErrs
+}
+
+// ValidateClusterCIDRConfigUpdate tests if an update to a ClusterCIDRConfig is valid.
+func ValidateClusterCIDRConfigUpdate(update, old *networking.ClusterCIDRConfig) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs, apivalidation.ValidateObjectMetaUpdate(&update.ObjectMeta, &old.ObjectMeta, field.NewPath("metadata"))...)
+	allErrs = append(allErrs, validateClusterCIDRConfigUpdateSpec(&update.Spec, &old.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+func validateClusterCIDRConfigUpdateSpec(update, old *networking.ClusterCIDRConfigSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(update.NodeSelector, old.NodeSelector, fldPath.Child("nodeSelector"))...)
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(update.PerNodeHostBits, old.PerNodeHostBits, fldPath.Child("perNodeHostBits"))...)
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(update.IPv4CIDR, old.IPv4CIDR, fldPath.Child("ipv4CIDR"))...)
+	allErrs = append(allErrs, apivalidation.ValidateImmutableField(update.IPv6CIDR, old.IPv6CIDR, fldPath.Child("ipv6CIDR"))...)
+
+	return allErrs
+}
```
@thockin if we don't allow Updates, do we need to add it to validation or is enough with denying it in the registry?

aojea commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -1982,3 +1982,186 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+		"valid IPv6 only ClusterCIDRConfig perNodeHostBits=100":  makeClusterCIDRConfigCustom(setIPv4CIDR(8, ""), setIPv6CIDR(100, "fd00:1:1::/16")),
+		"valid IPv4 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv6CIDR(8, "")),
+		"valid DualStack ClusterCIDRConfig with no NodeSelector": makeClusterCIDRConfigCustom(setNodeSelector(nil)),
+	}
```
missing edge cases on minimum and max PerNodeHostBits

aojea replied:
oh, yeah, it is missing that `10.1.0.0/16` with `16`  and `5` PerNodeHostBits is ok

sarveshr7 replied:
Done

aojea commented on pkg/apis/networking/validation/validation_test.go:
```diff
@@ -1982,3 +1982,186 @@ func TestValidateIngressStatusUpdate(t *testing.T) {
+
+	successCases := map[string]*networking.ClusterCIDRConfig{
+		"valid DualStack ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(),
+		"valid IPv6 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv4CIDR(8, "")),
```
```suggestion
		"valid IPv4 only ClusterCIDRConfig":                      makeClusterCIDRConfigCustom(setIPv4CIDR(8, "")),
```
??

aojea commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -249,6 +249,7 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			// used for pod eviction
 			rbacv1helpers.NewRule("update").Groups(legacyGroup).Resources("pods/status").RuleOrDie(),
 			rbacv1helpers.NewRule("list", "delete").Groups(legacyGroup).Resources("pods").RuleOrDie(),
+			rbacv1helpers.NewRule("get", "list").Groups(networkingGroup).Resources("clustercidrconfigs").RuleOrDie(),
```
is this controller consuming this API?

aojea replied:
nevermind

https://github.com/kubernetes/kubernetes/blob/9187b343e17db2bf8f3470cd6e1ef7f661814c15/cmd/kube-controller-manager/app/core.go#L101-L110

aojea commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,94 @@
+	// +optional
+	Spec ClusterCIDRConfigSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status is the current state of the ClusterCIDRConfig.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+	// +optional
+	Status ClusterCIDRConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
```
@thockin do we need status if we don't use it?

(I have the same question in the internal types)

sarveshr7 replied:
Removed

aojea commented:
We've talked offline, some of the outcome of the conversation between Tim, Sarvesh and I are

> Could we drop "Config" ?
> IPv4 vs IPv4CIDR

sarveshr7 committed:
Add describer and printer for ClusterCIDR API

sarveshr7 committed:
Introduce networking/v1alpha1 api, ClusterCIDR type

Introduce networking/v1alpha1 api group.

Add `ClusterCIDR` type to networking/v1alpha1 api group, this type
will enable the NodeIPAM controller to support multiple ClusterCIDRs.

sarveshr7 commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -370,3 +370,62 @@ func ScopedResourceSelectorRequirementsAsSelector(ssr v1.ScopedResourceSelectorR
+	selector := labels.NewSelector()
+	selector = selector.Add(requirements...)
+	return selector, nil
+}
```
NodeSelectorRequirementsAsSelector method is defined for `core.NodeSelectorRequirement` object however we need to use `core.v1.NodeSelectorRequirement`, also I wanted a method similar to LabelSelectorAsSelector which would take in the NodeSelector as input and return the selector as output, without iterating through the requirements everytime

Searched for the usage of `NodeSelectorRequirementsAsSelector` throughout the code, however it doesn't seem to be used anywhere

sarveshr7 commented:
/retest

sarveshr7 committed:
Auto generate code for ClusterCIDR API

thockin reviewed (commented):
Mostly minor things in the API PR.  Will try to look at IMPL in the AM

aojea commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -249,6 +249,7 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			// used for pod eviction
 			rbacv1helpers.NewRule("update").Groups(legacyGroup).Resources("pods/status").RuleOrDie(),
 			rbacv1helpers.NewRule("list", "delete").Groups(legacyGroup).Resources("pods").RuleOrDie(),
+			rbacv1helpers.NewRule("get", "list").Groups(networkingGroup).Resources("clustercidrs").RuleOrDie(),
```
I've done the same question before

https://github.com/kubernetes/kubernetes/blob/9187b343e17db2bf8f3470cd6e1ef7f661814c15/cmd/kube-controller-manager/app/core.go#L101-L110

the nodeIpamController identifies itself as "node-controller"

thockin replied:
I see, thanks

sarveshr7 commented on staging/src/k8s.io/kubectl/pkg/describe/describe.go:
```diff
@@ -2853,6 +2854,63 @@ func (i *IngressClassDescriber) describeIngressClassV1(ic *networkingv1.IngressC
+func (c *ClusterCIDRDescriber) Describe(namespace, name string, describerSettings DescriberSettings) (string, error) {
+	var events *corev1.EventList
+
+	cccV1alpha1, err := c.client.NetworkingV1alpha1().ClusterCIDRs().Get(context.TODO(), name, metav1.GetOptions{})
```
Done

sarveshr7 commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,90 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+
+	_, cidr, err := netutils.ParseCIDRSloppy(configCIDR)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("cidr"), cidr, fmt.Sprintf("must be a valid CIDR: %s", configCIDR)))
```
Done

sarveshr7 commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,90 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+		} else {
+			allErrs = append(allErrs, validatePerNodeHostBits(spec.IPv4, spec.PerNodeHostBits, 32, fldPath)...)
+		}
+
```
removed

sarveshr7 commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,90 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+
+	// Validate if CIDR is configured for at least one IP Family(IPv4/IPv6).
+	if spec.IPv4 == "" && spec.IPv6 == "" {
+		allErrs = append(allErrs, field.Required(fldPath, "one or both of `ipv4` and `ipv6` must be configured"))
```
Done

sarveshr7 commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -602,3 +602,90 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+var ValidateClusterCIDRName = apimachineryvalidation.NameIsDNSLabel
+
+// ValidateClusterCIDR validates a ClusterCIDR.
+func ValidateClusterCIDR(ccc *networking.ClusterCIDR) field.ErrorList {
```
Done

sarveshr7 commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -370,3 +370,62 @@ func ScopedResourceSelectorRequirementsAsSelector(ssr v1.ScopedResourceSelectorR
+// implements labels.Selector
+// Note: This function should be kept in sync with the selector methods in
+// pkg/labels/selector.go
+func NodeSelectorAsSelector(ns *v1.NodeSelector) (labels.Selector, error) {
```
It is used in the MultiCIDRRangeAllocator

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+
+	// IPv4 defines an IPv4 IP block in CIDR notation(e.g. "10.0.0.0/8").
+	// This field is immutable.
+	// +optional
```
Atleast one is necessary, documented

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+	// A subnet mask determines how much of the address is used for network bits
+	// and host bits. For example and IPv4 address of 192.168.0.0/24, splits the
+	// address into 24 bits for the network portion and 8 bits for the host portion.
+	// For a /24 mask for IPv4 or a /120 for IPv6, configure PerNodeHostBits=8
```
Done

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+	// and host bits. For example and IPv4 address of 192.168.0.0/24, splits the
+	// address into 24 bits for the network portion and 8 bits for the host portion.
+	// For a /24 mask for IPv4 or a /120 for IPv6, configure PerNodeHostBits=8
+	// Minimum value for PerNodeHostBits is 4.
```
Done

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+	// A subnet mask determines how much of the address is used for network bits
+	// and host bits. For example and IPv4 address of 192.168.0.0/24, splits the
+	// address into 24 bits for the network portion and 8 bits for the host portion.
+	// For a /24 mask for IPv4 or a /120 for IPv6, configure PerNodeHostBits=8
```
Done

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+
+	// PerNodeHostBits defines the number of host bits to be configured per node.
+	// A subnet mask determines how much of the address is used for network bits
+	// and host bits. For example and IPv4 address of 192.168.0.0/24, splits the
```
Done

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+// ClusterCIDRSpec defines the desired state of ClusterCIDR.
+type ClusterCIDRSpec struct {
+	// NodeSelector defines which nodes the config is applicable to.
+	// An empty or nil NodeSelector functions as a default that applies to all nodes.
```
Done

sarveshr7 commented on staging/src/k8s.io/api/networking/v1alpha1/types.go:
```diff
@@ -0,0 +1,105 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.25
+
+// ClusterCIDR is the Schema for the ClusterCIDR API.
```
Done

sarveshr7 commented:
/test pull-kubernetes-e2e-ubuntu-gce-network-policies

cici37 commented:
Hello 👋, 1.25 Release Lead here.

The [exception request](https://groups.google.com/g/kubernetes-sig-release/c/U8etwqKzy64) is approved and your updated deadline to make any changes to your PR is **3:30 AM PST Monday 8th August 2022**. Thank you!

/milestone v1.25

thockin commented:
Let's focus the review on https://github.com/kubernetes/kubernetes/pull/109090 which includes these commits, so we can merge atomically or not at all.  Fair?

sarveshr7 commented:
Sounds good! Had separated the PRs to make the review easier, as the API review is done and the commits are present in https://github.com/kubernetes/kubernetes/pull/109090, we can merge atomically.
